### ISSUE TITLE
cnetedit - removing cubes from a control network

### ISSUE THREAD

**Issue by ascbot** [DOI-USGS/ISIS3#1990]
---

Author Name: **Tammy Becker** (Tammy Becker)

Original Date: 2015-04-01
Original Assignee: Kristin Berry

---

cnetedit 'cubelist' feature has a bug.  This feature is intended to allow the user to supply a list of filenames to edit in the control network. 
The most common 'edit' is to remove the cubes from the network (i.e., 'remove the measures' associated with the list of cube filenames from the network by setting to ignore then deleting the ignored)...unfortunately, when this feature is used, not only do the measures with filenames from cubelist get removed, but the entire point with all the other measures not included in the cubelist are removed.  This is irregardless of setting the other flags for 'preserve', retain_reference and ignoreall.

The current work around is to extract the filenames we want to keep in the network using cnetextract  :-/

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Moses Milazzo** (Moses Milazzo)
Original Date: 2015-04-02T16:51:44Z

---

Added to FY15Q3 as top priority fix.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Lynn Weller** (Lynn Weller)
Original Date: 2015-04-02T22:17:45Z

---

I have a small set of data that demonstrates this problem under /work/users/lweller/Isis3Tests/CnetEdit/CubeList/.

command line:

  cnetedit cnet=autoseed_referenceSet.net onet=RemoveCube.net cubelist=remove_cube.lis log=log_RemoveCube

The resulting network has one point in it but it should have 2.  

In order to get the correct result, one could use cnetextract as Tammy has pointed out or the user can make a list of measures associated with the cube that is to be removed and pass that to cnetedit instead (this takes an extra step).  This works as demonstrated in the command and output below:

  cnetedit cnet=autoseed_referenceSet.net onet=RemoveMeasures.net measurelist=remove_measures.lis log=log_RemoveMeasures

I highly recommend taking a look at the network in qnet and/or qmos (with points displayed) and convince yourself of what should happen and compare that against what is happening.  I have generated a number of output files from cnetstats to also shed light on the network so it is hopefully clearer to the developer what is going on.  Of course Tammy and I can always sit with someone and point things out if it's not otherwise clear.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-06-04T00:21:21Z

---

The commands in #2 above both produce identical output control nets using the most recent Isis beta. Could not reproduce using given examples. Does this problem still show up in other cases? (Or do I not understand what the problem is?)

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Lynn Weller** (Lynn Weller)
Original Date: 2015-06-04T17:29:55Z

---

Looking into this right now, but you're right, under beta both methods produce the same results which is not what Tammy and I had been experincing.  Tammy needs to have a look as well.  One of us will let you know how it goes.  We're both in the middle of things at the moment so it may be a day or two before we can revisit this to understand what's going on.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-06-04T18:08:54Z

---

Thanks for the update and for letting me know that it might be a little while. I definitely have plenty of other things to work on! :)

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-06-30T18:33:10Z

---

Currently cnetedit sets the whole point to ignore whenever the reference measure is set to ignore (making IGNOREALL not very useful.) The decided-upon fix is to get rid of this behavior: the whole point is not ignored when the reference measure is ignored unless IGNOREALL=true.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-07-01T00:29:04Z

---

As far as I can tell, this now preforms the correct behavior, but I did only test a few different cases. Please let me know if anything seems incorrect. 

Also: The output log looks a little bit different after this change. Everything made sense to me, but please let me know if it does not to you or if you would like any changes.

To test setisis /work/projects/isis/latest/m02238/isis

Thanks!

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Tammy Becker** (Tammy Becker)
Original Date: 2015-07-01T02:31:41Z

---

I've tested this with a netfile where the cube that is 'removed' is a reference image for points.  It works now as expected.

One more thing, unfortunately, when removing a cube that is a reference that leaves a point with only a single measure, nothing happens to this point.  it should probably be set to ignore, unless preserve is set to true..then the point should be left valid.

Let's have a group discussion when you can to discuss this last tidbit.

my test area is /work/users/tbecker/IsisTesting/M02238_cnetedit/

proc.csh contains my commands and input/output net files

Thank you!

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-07-15T23:22:40Z

---

Your 2nd test in proc.csh now works. 13,14, and 15 are kept if preserve = true and deleted otherwise. 

The first test has identical control nets for input and output. Hopefully this is expected? (I didn't see any Ignored points or measures in the input control net.)

The 3rd test works as I would expect it to-- if there's no deletes, we never even get to the point where preserve=true makes a difference.

The 4th test with/without ignoreAll=true works as I would expect: When ignoreAll = true, if a reference measure is Ignored, the whole point and all its measures are Ignored.

I wasn't sure what you were testing in the last one, so that's up to you :)

To test some more: setisis /work/projects/isis/latest/m02238/isis

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Tammy Becker** (Tammy Becker)
Original Date: 2015-07-29T00:49:02Z

---

It looks like all the options work.  

I need to talk to you about the state of remaining single measures, regardless of the options, they are deleted (with delete=yes) or they seem to always remain valid and never set to 'ignore' which is what I thought the 'preserve=true' would do...identify the single measures and with delete=no, preserve them but they are set to ignore...either the code doesn't allow this (if this is the case, we might need a separate ticket? because we are heading beyond the purpose of this ticket)...or I am misunderstanding, which is entirely possible.

thank you!

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Tammy Becker** (Tammy Becker)
Original Date: 2015-07-29T19:07:27Z

---

Regarding the single measure issue; confirmed that the delete=yes & preserve=yes retains the single measures/points (valid).  This is the ultimately the important functionality.   If we want to set the single measures/points to ignore then it will be a separate feature request.   For now, the preserve=yes is sufficient.  Another way to track single measures is through cnetcheck.    Lynn and Tammy agree that this is all good for now.

Will close this post!

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-08-03T19:54:27Z

---

Lynn found a problem with this. 

Using a measurelist with all default options (including DELETE=YES,) the measures in the measurelist are just ignored, rather than deleted. I'm trying to track this down.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-08-03T20:25:37Z

---

The new problem is that reference measures never get ignored.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-08-04T20:17:55Z

---

It's ready for testing again with
setisis /work/projects/isis/latest/m02238/isis

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Lynn Weller** (Lynn Weller)
Original Date: 2015-08-04T21:03:03Z

---

Tested this on measurelist, cubelist and pointlist with delete set to true or false for each case and got expected results.  Have not tested the other options.  I will check using a deffile since I use that functionality a lot, but I won't get to this immediately today.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Kristin Berry** (Kristin Berry)
Original Date: 2015-08-11T21:03:49Z

---

Tammy will update documentation to warn of non-intuitive results that occur when RETAIN_REF=yes, but PRESERVE=no (And maybe more cases?)

With RETAIN_REF=yes, PRESERVE=no, and DELETE=yes, reference measures aren't directly ignored, but if they are the only remaining measure that a point has, they will be deleted from the control network along with the point.

**Comment by ascbot** [comment]
---

Original Redmine Comment
Author Name: **Tammy Becker** (Tammy Becker)
Original Date: 2015-08-11T23:32:37Z

---

Did a lot of testing in /work/users/tbecker/IsisTesting/M02238_cnetedit/proc.csh for measurelist, checkvalid and pointlist.  
We updated documentation for correctness.
